crates/ruff_linter/src/rules/ruff/rules/sequence_sorting.rs

```diff
@@ -0,0 +1,878 @@
+    elements: &[&str],
+    locator: &Locator,
+    sorting_style: &SortingStyle,
+) -> String {
+    assert_eq!(elts.len(), elements.len());
+    let (opening_paren, closing_paren) = kind.surrounding_parens(locator.contents());
+    assert!(
+        elements.len() >= 2,
+        "A sequence with < 2 elements cannot be unsorted"
+    );
+    let last_item_index = elements.len() - 1;
+    let mut result = String::from(opening_paren);
+
+    let mut element_pairs = elements.iter().zip(elts).collect_vec();
+    element_pairs.sort_by(|(elem1, _), (elem2, _)| sorting_style.compare(elem1, elem2));
+    // We grab the original source-code ranges using `locator.slice()`
+    // rather than using the expression generator, as this approach allows
+    // us to easily preserve stylistic choices in the original source code
+    // such as whether double or single quotes were used.
+    for (i, (_, elt)) in element_pairs.iter().enumerate() {
+        result.push_str(locator.slice(elt));
+        if i < last_item_index {
+            result.push_str(", ");
+        }
+    }
+
+    result.push_str(closing_paren);
+    result
+}
+
+/// An enumeration of the possible conclusions we could come to
+/// regarding the ordering of the elements in a display of string literals
+#[derive(Debug, is_macro::Is)]
+pub(super) enum SortClassification<'a> {
+    /// It's a display of string literals that is already sorted
+    Sorted,
+    /// It's an unsorted display of string literals,
+    /// but we wouldn't be able to autofix it
+    UnsortedButUnfixable,
+    /// It's an unsorted display of string literals,
+    /// and it's possible we could generate a fix for it
+    UnsortedAndMaybeFixable { items: Vec<&'a str> },
+    /// The display contains one or more items that are not string
+    /// literals.
+    NotAListOfStringLiterals,
+}
+
+impl<'a> SortClassification<'a> {
+    pub(super) fn of_elements(elements: &'a [ast::Expr], sorting_style: &SortingStyle) -> Self {
+        let Some((first, rest @ [_, ..])) = elements.split_first() else {
+            return Self::Sorted;
+        };
+        let Some(string_node) = first.as_string_literal_expr() else {
+            return Self::NotAListOfStringLiterals;
+        };
+        let mut this = string_node.value.to_str();
+
+        for expr in rest {
+            let Some(string_node) = expr.as_string_literal_expr() else {
+                return Self::NotAListOfStringLiterals;
+            };
+            let next = string_node.value.to_str();
+            if sorting_style.compare(next, this).is_lt() {
+                let mut items = Vec::with_capacity(elements.len());
+                for expr in elements {
+                    let Some(string_node) = expr.as_string_literal_expr() else {
+                        return Self::NotAListOfStringLiterals;
+                    };
+                    if string_node.value.is_implicit_concatenated() {
+                        return Self::UnsortedButUnfixable;
```

Comment:
Could we possibly return `Self::UnsortedButUnfixable` when a later element in `elements` is not a string literal? Like what if the first two elements are string literals, and the second element is implicitly concatenated, and then the third element is not a string. (I suspect we should always return `NotAListOfStringLiterals` if any element is not a string, but I'm wondering if that's the case with the coed as-written.)

Reply:
Hmm, good point

Reply:
Fixed in https://github.com/astral-sh/ruff/pull/9564/commits/606d85a74cfb8be467be0882fe94081c08463542, and I added a test!